cmd/init-osm-controller/init-osm-controller.go

```diff
@@ -0,0 +1,110 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"os"
+	"time"
+
+	"github.com/spf13/pflag"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/openservicemesh/osm/pkg/apis/config/v1alpha1"
+	"github.com/openservicemesh/osm/pkg/cli"
+	configClientset "github.com/openservicemesh/osm/pkg/gen/client/config/clientset/versioned"
+	"github.com/openservicemesh/osm/pkg/logger"
+	"github.com/openservicemesh/osm/pkg/version"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const meshConfigName = "osm-mesh-config"
+
+var settings = cli.New()
+
+var (
+	flags = pflag.NewFlagSet("init-osm-controller", pflag.ExitOnError)
+	log   = logger.New("osm-controller/main")
+
+	meshName           string // An ID that uniquely identifies an OSM instance
+	kubeConfigFilePath string
+	osmNamespace       string
+
+	fetchMeshConfigMaxRetry      = 5
+	fetchMeshConfigRetryInterval = 5
+)
+
+func init() {
+	if err := logger.SetLogLevel("info"); err != nil {
+		log.Fatal().Err(err).Msg("Error setting log level")
+	}
+
+	flags.StringVar(&meshName, "mesh-name", "", "OSM mesh name")
+	flags.StringVar(&kubeConfigFilePath, "kubeconfig", "", "Path to Kubernetes config file.")
+	flags.StringVar(&osmNamespace, "osm-namespace", "", "Namespace to which OSM belongs to.")
+}
+
+func parseFlags() error {
+	if err := flags.Parse(os.Args); err != nil {
+		return err
+	}
+	_ = flag.CommandLine.Parse([]string{})
+	return nil
+}
+
+func main() {
+	log.Info().Msgf("Starting init-osm-controller %s; %s; %s", version.Version, version.GitCommit, version.BuildDate)
+
+	if err := parseFlags(); err != nil {
+		log.Fatal().Err(err).Msg("Failed parsing cmd line arguments")
+	}
+
+	// Initialize kube config and client
+	kubeConfig, err := clientcmd.BuildConfigFromFlags("", kubeConfigFilePath)
+	if err != nil {
+		log.Fatal().Err(err).Msgf("Failed creating kube config (kubeconfig=%s)", kubeConfigFilePath)
+	}
+
+	configClient, err := configClientset.NewForConfig(kubeConfig)
+	if err != nil {
+		log.Fatal().Err(err).Msgf("Could not access Kubernetes cluster, check kubeconfig.")
+		return
+	}
+
+	var retry int
+	for retry = 0; retry < fetchMeshConfigMaxRetry; retry++ {
```

Comment:
This seems a bit hacky. Why is the retry required to create the default config? The CRD should always be present prior to the init container starting up, so the creation of the default config should never fail. Is there something I am missing?

Reply:
In this design, the purpose of init container is to create the default config, by itself. Because the default MeshConfig is not stored in helm charts.

The init container here will create the default MeshConfig if it does not exist.

Thus, if multiple osm-controller pod exist, they will race to create this default one. The retry strategy is trying to ensure only one pod creates it, and others wait for it to come up, and not overload the api server.

One assumption I made here, which I'm not sure, is the api-server will provide the mutex or stuff to ensure only one MeshConfig will be processed at a time. But even if it's not, since the MeshConfig will have the same name and content, I think it should be fine.

Reply:
I would suggest tweaking the implementation and make it similar to the code described in https://github.com/openservicemesh/osm/pull/3222#discussion_r620641915

Reply:
I will explore whether that really solves the race issue. But for the coding style, sure I will update.

Reply:
It does solve the race issue between multiple osm-controller's, which is why I am suggesting to change it :-)